flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayDistinctFunction.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_DISTINCT}. */
+@Internal
+public class ArrayDistinctFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+
+    public ArrayDistinctFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_DISTINCT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType();
+        elementGetter = ArrayData.createElementGetter(dataType.getLogicalType());
+    }
+
+    public @Nullable ArrayData eval(ArrayData haystack) {
+        try {
+            if (haystack == null) {
+                return null;
+            }
+            Set set = new LinkedHashSet<>();
+            final int size = haystack.size();
+            for (int pos = 0; pos < size; pos++) {
+                final Object element = elementGetter.getElementOrNull(haystack, pos);
+                set.add(element);
```

Comment:
@snuyanzin We need to be careful with equality of `RowData`, `MapData` and others. I'm not saying that this implementation is wrong, maybe it works for most cases. But I'm also not sure whether there is a hidden bug.

If you look at `BinaryRowData`, it compares binary sections. Which means that equality is not based on SQL semantics. This is actually why I added `SpecializedFunction` with a `ExpressionEvaluatorFactory`. Only with this piece it is now possible to call the SQL compliant equals function which can be used to assemble a proper distinct for all data types.

Reply:
I don’t have the time to find a bug, but I just wanted to point out the existence of the expression evaluator to get access to primitive functions such as CAST or EQUALS. CC @luoyuxia

Reply:
hi @twalthr thanks for reminding us. but i have a question with array_contains. why element and needle type is not null. the design makes develop confused. and it cause a bug, you can see here https://github.com/apache/flink/pull/21993

        equalityEvaluator =
                context.createEvaluator(
                        $("element").isEqual($("needle")),
                        DataTypes.BOOLEAN(),
                        DataTypes.FIELD("element", needleDataType.notNull().toInternal()),
                        DataTypes.FIELD("needle", needleDataType.notNull().toInternal()));